### PR TITLE
Dynamic version of id set

### PR DESCRIPTION
**Pull request by MosheEichler**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/32099

## Description
Create a dynamic id_set to for find_dependencies UT

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: TestSuite/integration.py [hunk 1]
```diff
@@ -56,13 +56,19 @@ def build(
...
 
         with open(suite_join_path(default_integration_dir, 'sample.py')) as code_file:
             code = str(code_file.read())
         with open(suite_join_path(default_integration_dir, 'sample.yml')) as yml_file:
             yml = yaml.safe_load(yml_file)
+            if name:
+                yml['name'] = name
+                yml['commonfields']['id'] = name
+            if commands:
+                for command in commands:
```
Comment (teizenman) [inline TestSuite/integration.py:70]: `commands` is a string according to the annotation above.
We should either accept it as a comma-separated value or as a list.

File: TestSuite/repo.py [hunk 2]
```diff
@@ -133,7 +133,7 @@ def setup_one_pack(self, name) -> Pack:
 
         indicator_type = pack.create_indicator_type(f'{name}_indicator-type')
         indicator_type.write_json({'id': f'{name} - indicator_type'})
-        indicator_type.update({'name': f'{name} - indicator_type'})
+        indicator_type.update({'details': f'{name} - indicator_type'})
```
Comment (teizenman) [inline TestSuite/repo.py:136]: Why did you change it?

**Review by teizenman — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py [hunk 3]
```diff
@@ -1,22 +1,231 @@
...
+    impossible_traveler.create_incident_field('Previous_Source_IP', {'id': 'incident_previoussourceip',
+                                                                     'name': 'Previous Source IP'})
+    impossible_traveler.create_incident_field('Sign_In_Date_Time', {'id': 'incident_signindatetime',
+                                                                    'name': 'Sign In Date Time'})
+    impossible_traveler.create_incident_field('Travel_Map_Link', {'id': 'incident_travelmaplink',
+                                                                  'name': 'Travel Map Link'})
+    impossible_traveler.create_incident_type('Impossible_Traveler', {'id': 'impossibletraveler',
+                                                                     'name': 'Impossible Traveler',
+                                                                     "playbookId": "Impossible Traveler",
+                                                                     'preProcessingScript': '', 'color': 'test'})
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:223]: This creation is very long, please consider splitting it to helper methods ( for each pack) or at least order in sections

**Review by reutshal — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py [hunk 4]
```diff
@@ -1,22 +1,271 @@
...
+    # Create a pack called 'CommonTypes' with 3 incident_fields 2 incident_types 5 indicator_fields 1 indicator_type.
+    common_types = repo.create_pack('CommonTypes')
+    ct_incident_field_ids = ['accountid', 'country', 'username']
+    ct_incident_field_names = ['Account Id', 'Country', 'Username']
+    ct_incident_type_ids = ['Network', 'Authentication']
+    ct_incident_type_names = ['Network', 'Authentication']
+    ct_indicator_field_ids = ['accounttype', 'adminname', 'tags', 'commontypes', 'adminemail']
+    ct_indicator_field_names = ['Account Type', 'Admin Name', 'Tags', 'Common Types', 'Admin Email']
+    for i in range(len(ct_incident_field_ids)):
+        create_a_pack_entity(common_types, 'incident_field', ct_incident_field_ids[i], ct_incident_field_names[i])
```
Comment (yaakovi) [inline demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:154]: for all similar loops
```suggestion
    for field_id, field_name in zip(ct_incident_field_ids, ct_incident_field_names):
        create_a_pack_entity(common_types, 'incident_field', field_id, field_name)
```

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py [hunk 5]
```diff
@@ -1,234 +1,289 @@
...
+    Given
+        - A Pack.
+
+    When
+        - add an entity to the pack.
+
+    Then
+        - Adds the entity to the pack with basic data.
+    """
+    if entity_type == 'script':
```
Comment (teizenman) [inline demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:29]: I would suggest creating an enum for it, it's always better to have pre-defined values in those cases.

File: TestSuite/integration.py [hunk 6]
```diff
@@ -56,16 +56,15 @@ def build(
         if image is not None:
             self.image.write_bytes(image)
 
-    def create_default_integration(self, name: str = '', commands: str = ''):
+    def create_default_integration(self, name: str = None, commands: List[str] = None):
```
Comment (teizenman) [inline TestSuite/integration.py:59]: Add docstrings to all "create default" functions, now that there is real logic happening.

**Review by teizenman — CHANGES_REQUESTED** [review]
Why did you merge all the test classes into one?

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py [hunk 7]
```diff
@@ -1,24 +1,294 @@
...
+    create_a_pack_entity(impossible_traveler, FileType.INCIDENT_FIELD, 'signindatetime' 'Sign In Date Time')
+    create_a_pack_entity(impossible_traveler, FileType.INCIDENT_FIELD, 'travelmaplink' 'Travel Map Link')
+    create_a_pack_entity(impossible_traveler, FileType.INCIDENT_TYPE, 'impossibletraveler' 'Impossible Traveler')
+
+    with ChangeCWD(repo.path):
+        ids = cis.IDSetCreator()
+        ids.create_id_set()
+
+
+def setup_module():
```
Comment (MosheEichler) [inline demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:288]: This function creates the content repo just for the current file and removes the repo after all UT are completed, It's something like this
    
    def setup_class(cls):
        create_content_repo()

    def teardown_class(cls):
        remove_content_repo()

but instead of creating and removing the content repo for each class it only removes after it finishes all the classes of the UT file (the UT file contains some TestClasses)

[link to pytest documentation](https://docs.pytest.org/en/6.2.x/xunit_setup.html)

File: TestSuite/integration.py (hunk 6 above)
Comment (teizenman) [reply TestSuite/integration.py:59]: That's the convention for UT docstrings, not for used functions.
Please change it.

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py [hunk 8]
```diff
@@ -1,24 +1,294 @@
...
+    elif entity_type == FileType.INTEGRATION:
+        pack.create_integration(entity_id).create_default_integration(entity_id, commands)
+    elif entity_type == FileType.PLAYBOOK:
+        pack.create_playbook(entity_id).create_default_playbook(entity_id)
+    elif entity_type == FileType.TEST_PLAYBOOK:
+        pack.create_test_playbook(entity_id).create_default_test_playbook(entity_id)
+    elif entity_type == FileType.CLASSIFIER:
+        content = {'id': entity_id, 'name': entity_name, 'transformer': '', 'keyTypeMap': {}, 'type': 'classification'}
+        pack.create_classifier(entity_id, content)
+    elif entity_type == FileType.LAYOUT:
```
Comment (teizenman) [inline demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:40]: Are you sure it's `layout` and not `layoutcontainer`?

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py (hunk 7 above)
Comment (teizenman) [reply demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:288]: Don't you need `teardown_module` as well?

**Review by teizenman — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py (hunk 8 above)
Comment (MosheEichler) [reply demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:40]: Yes. 'layoutcontainer' has no key 'kind' instead it has the 'group' key

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py (hunk 7 above)
Comment (MosheEichler) [reply demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:288]: It seems not. The files are **automatically** deleted after all tests are completed

File: TestSuite/integration.py [hunk 9]
```diff
@@ -57,13 +57,14 @@ def build(
...
 
     def create_default_integration(self, name: str = None, commands: List[str] = None):
-        """
-        Given:
-            - A pack with an empty integration file
-        When
-            - Creating a default integration entity with/out a specific name and ID and with/out additional commands
-        Then
-            - Checks if there is a specific name for the integration (default is "Sample") and creates the integration
+        """Creates a new integration with/out a specific name and ID with basic data
```
Comment (reutshal) [inline TestSuite/integration.py:60]:
```suggestion
        """Creates a new integration with basic data.
```

File: TestSuite/integration.py [hunk 10]
```diff
@@ -57,13 +57,14 @@ def build(
...
+
+        :type name: ``Optional[str]``
+        :param name: The name and ID of the new playbook, default is "sample playbook".
```
Comment (reutshal) [inline TestSuite/integration.py:63]:
```suggestion
        :param name: The name and ID of the new integration, default is "sample playbook".
```

Also fix the default here

File: TestSuite/integration.py [hunk 11]
```diff
@@ -57,13 +57,14 @@ def build(
...
+
+        :type commands: ``Optional[List[str]]``
+        :param name: List of additional commands to add to the integration.
```
Comment (reutshal) [inline TestSuite/integration.py:66]:
```suggestion
        :param commands: List of additional commands to add to the integration.
```

File: TestSuite/playbook.py [hunk 12]
```diff
@@ -41,14 +41,11 @@ def build(
...
     def create_default_playbook(self, name: str = None):
-        """
-        Given:
-            - A pack with an empty playbook file
-        When
-            - Creating a default playbook entity with/out a specific name and ID
-        Then
-            - Checks if there is a specific name for the playbook
-             (default is "sample playbook") and creates the playbook
+        """Creates a new playbook with/out a specific name and ID with basic data
```
Comment (reutshal) [inline TestSuite/playbook.py:44]:
```suggestion
        """Creates a new playbook with basic data.
```

File: TestSuite/script.py [hunk 13]
```diff
@@ -14,13 +14,11 @@ def __init__(self, tmpdir: Path, name, repo, create_unified=False):
...
 
     def create_default_script(self, name: str = None):
-        """
-        Given:
-            - A pack with an empty script file
-        When
-            - Creating a default script entity with/out a specific name and ID
-        Then
-            - Checks if there is a specific name for the script (default is "sample_script") and creates the script
+        """Creates a new script with/out a specific name and ID with basic data
```
Comment (reutshal) [inline TestSuite/script.py:17]:
```suggestion
        """Creates a new script with basic data.
```

**Review by reutshal — CHANGES_REQUESTED** [review]

File: TestSuite/integration.py (hunk 6 above)
Comment (reutshal) [reply TestSuite/integration.py:59]: @MosheEichler Please use google docstring: 
image

you can read about it here
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py [hunk 14]
```diff
@@ -1,24 +1,294 @@
...
+    prisma_cloud_compute_scripts = ['PrismaCloudComputeParseAuditAlert',
+                                    'PrismaCloudComputeParseCloudDiscoveryAlert',
+                                    'PrismaCloudComputeParseComplianceAlert',
+                                    'PrismaCloudComputeParseVulnerabilityAlert']
+    for script in prisma_cloud_compute_scripts:
+        create_a_pack_entity(prisma_cloud_compute, FileType.SCRIPT, script)
+    create_a_pack_entity(prisma_cloud_compute, FileType.INCIDENT_TYPE, 'Prisma Cloud Compute Cloud Discovery',
+                         'Prisma Cloud Compute Cloud Discovery')
+
+    # Create a pack called 'PrismaCloudCompute' with 1 playbook.
```
Comment (reutshal) [inline demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:85]:
```suggestion
    # Create a pack called 'Expanse' with 1 playbook.
```

**Review by reutshal — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py (hunk 8 above)
Comment (teizenman) [reply demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:40]: So, what about layoutcontainer then?

Comment (MosheEichler) [reply demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:40]: In the UT there is no a pack depends on another pack by a layoutcontainer yet
But for the future you are right we should be able to create a layoutcontainer also
So I added that

File: demisto_sdk/commands/common/constants.py [hunk 15]
```diff
@@ -73,6 +73,7 @@ class FileType(Enum):
     LAYOUTS_CONTAINER = 'layoutscontainer'
     DASHBOARD = 'dashboard'
     INCIDENT_TYPE = 'incidenttype'
+    INDICATOR_TYPE = 'indicatortype'
```
Comment (yaakovi) [inline demisto_sdk/commands/common/constants.py:76]: this is `reputation`

File: TestSuite/script.py [hunk 16]
```diff
@@ -13,13 +13,22 @@ class Script(Integration):
     def __init__(self, tmpdir: Path, name, repo, create_unified=False):
         super().__init__(tmpdir, name, repo, create_unified)
 
-    def create_default_script(self):
+    def create_default_script(self, name: str = None):
+        """Creates a new script with basic data.
+
+        Args:
+            name: The name and ID of the new script, default is "sample_script".
```
Comment (yaakovi) [inline TestSuite/script.py:20]:
```suggestion
    def create_default_script(self, name: str = 'sample_script'):
        """Creates a new script with basic data.

        Args:
            name: The name and ID of the new script, default is "sample_script".
```

File: TestSuite/script.py [hunk 17]
```diff
@@ -13,13 +13,22 @@ class Script(Integration):
...
+        """
+
         default_script_dir = 'assets/default_script'
 
         with open(suite_join_path(default_script_dir, 'sample_script.py')) as code_file:
             code = str(code_file.read())
         with open(suite_join_path(default_script_dir, 'sample_script.yml')) as yml_file:
             yml = yaml.load(yml_file, Loader=yaml.FullLoader)
+            if name:
+                yml['name'] = yml['commonfields']['id'] = name
```
Comment (yaakovi) [inline TestSuite/script.py:31]:
```suggestion
            yml['name'] = yml['commonfields']['id'] = name
```

File: TestSuite/playbook.py [hunk 18]
```diff
@@ -40,14 +40,24 @@ def build(
         if not self.is_test_playbook and readme is not None:
             self.readme.write(readme)
 
-    def create_default_playbook(self):
+    def create_default_playbook(self, name: str = None):
+        """Creates a new playbook with basic data.
+
+        Args:
+            name: The name and ID of the new playbook, default is "sample playbook".
```
Comment (yaakovi) [inline TestSuite/playbook.py:47]:
```suggestion
    def create_default_playbook(self, name: str = 'sample playbook'):
        """Creates a new playbook with basic data.

        Args:
            name: The name and ID of the new playbook, default is "sample playbook".
```

File: TestSuite/playbook.py [hunk 19]
```diff
@@ -40,14 +40,24 @@ def build(
...
+    def create_default_test_playbook(self, name: str = None):
         default_test_playbook_dir = 'assets/default_playbook'
-        with open(suite_join_path(default_test_playbook_dir, 'playbook-sample.yml')) as yml:
-            self.build(
-                yml=yaml.safe_load(yml)
-            )
+        with open(suite_join_path(default_test_playbook_dir, 'playbook-sample.yml')) as yml_file:
+            yml = yaml.safe_load(yml_file)
+            if name:
+                yml['id'] = yml['name'] = name
```
Comment (yaakovi) [inline TestSuite/playbook.py:62]:
```suggestion
            yml['id'] = yml['name'] = name
```

File: TestSuite/playbook.py [hunk 20]
```diff
@@ -40,14 +40,24 @@ def build(
...
+            name: The name and ID of the new playbook, default is "sample playbook".
+
+        """
         default_playbook_dir = 'assets/default_playbook'
-        with open(suite_join_path(default_playbook_dir, 'playbook-sample.yml')) as yml:
-            self.build(yml=yaml.safe_load(yml))
+        with open(suite_join_path(default_playbook_dir, 'playbook-sample.yml')) as yml_file:
+            yml = yaml.safe_load(yml_file)
+            if name:
+                yml['id'] = yml['name'] = name
```
Comment (yaakovi) [inline TestSuite/playbook.py:54]:
```suggestion
            yml['id'] = yml['name'] = name
```

File: TestSuite/playbook.py [hunk 21]
```diff
@@ -40,14 +40,24 @@ def build(
...
+            self.build(yml=yml)
 
-    def create_default_test_playbook(self):
+    def create_default_test_playbook(self, name: str = None):
```
Comment (yaakovi) [inline TestSuite/playbook.py:57]: replace with the current default value.
```suggestion
    def create_default_test_playbook(self, name: str = 'SamplePlaybookTest):
```

File: TestSuite/integration.py (hunk 6 above)
Comment (yaakovi) [reply TestSuite/integration.py:59]:
```suggestion
    def create_default_integration(self, name: str = 'Sample', commands: List[str] = None):
```

File: TestSuite/integration.py [hunk 22]
```diff
@@ -56,13 +56,25 @@ def build(
...
+
+        """
         default_integration_dir = 'assets/default_integration'
 
         with open(suite_join_path(default_integration_dir, 'sample.py')) as code_file:
             code = str(code_file.read())
         with open(suite_join_path(default_integration_dir, 'sample.yml')) as yml_file:
             yml = yaml.safe_load(yml_file)
+            if name:
+                yml['name'] = yml['commonfields']['id'] = name
```
Comment (yaakovi) [inline TestSuite/integration.py:74]:
```suggestion
            yml['name'] = yml['commonfields']['id'] = name
```

File: demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py [hunk 23]
```diff
@@ -1038,10 +1279,10 @@ def test_collect_incident_layout_dependencies(self, id_set):
 
         assert IsEqualFunctions.is_sets_equal(found_result, expected_result)
 
-    def test_collect_indicator_layout_dependencies(self, id_set):
+    def test_collect_indicaotr_layouts_dependencies(self, id_set):
```
Comment (yaakovi) [inline demisto_sdk/commands/find_dependencies/tests/find_dependencies_test.py:1282]:
```suggestion
    def test_collect_indicator_layouts_dependencies(self, id_set):
```

**Review by yaakovi — APPROVED** [review]

File: TestSuite/playbook.py [hunk 24]
```diff
@@ -40,24 +40,22 @@ def build(
...
         default_playbook_dir = 'assets/default_playbook'
         with open(suite_join_path(default_playbook_dir, 'playbook-sample.yml')) as yml_file:
             yml = yaml.safe_load(yml_file)
-            if name:
-                yml['id'] = yml['name'] = name
+            yml['id'] = yml['name'] = name
             self.build(yml=yml)
 
-    def create_default_test_playbook(self, name: str = None):
+    def create_default_test_playbook(self, name: str = 'SamplePlaybookTest'):
```
Comment (reutshal) [inline TestSuite/playbook.py:56]: Please keep the default names consistent.
We have "Sample", "sample playbook" and "SamplePlaybookTest".
Lets keep them in the same format.

File: TestSuite/script.py [hunk 25]
```diff
@@ -13,11 +13,11 @@ class Script(Integration):
     def __init__(self, tmpdir: Path, name, repo, create_unified=False):
         super().__init__(tmpdir, name, repo, create_unified)
 
-    def create_default_script(self, name: str = None):
+    def create_default_script(self, name: str = 'sample_script'):
```
Comment (reutshal) [inline TestSuite/script.py:16]: Same here

**Review by reutshal — APPROVED** [review]

**Review by teizenman — APPROVED** [review]
